### PR TITLE
fpmsyncd: Re-allocating link cache when interface name cannot be resolved

### PR DESCRIPTION
**Pull request by stcheng**
- This happens when the host interfaces change and then we need to update
  the link cache.

### CODE REVIEW

File: fpmsyncd/routesync.cpp [hunk 1]
```diff
@@ -110,6 +110,10 @@ void RouteSync::onMsg(int nlmsg_type, struct nl_object *obj)
         if (m_link_cache)
         {
             rtnl_link_i2name(m_link_cache, ifindex, addrStr, MAX_ADDR_SIZE);
+            if (!strlen(addrStr))
+            {
+                rtnl_link_alloc_cache(m_nl_sock, AF_UNSPEC, &m_link_cache);
+            }
             ifindexes += addrStr;
```
Comment (lguohan) [inline fpmsyncd/routesync.cpp]: ifindexes rename to ifnames

addrStr rename to ifname

File: fpmsyncd/routesync.cpp [hunk 2]
```diff
@@ -110,6 +110,10 @@ void RouteSync::onMsg(int nlmsg_type, struct nl_object *obj)
         if (m_link_cache)
         {
             rtnl_link_i2name(m_link_cache, ifindex, addrStr, MAX_ADDR_SIZE);
+            if (!strlen(addrStr))
+            {
+                rtnl_link_alloc_cache(m_nl_sock, AF_UNSPEC, &m_link_cache);
+            }
```
Comment (lguohan) [inline fpmsyncd/routesync.cpp]: after you refresh the cache, do you need to resolve the name again?

**Comment by stcheng** [comment]
Updated.

File: fpmsyncd/routesync.cpp [hunk 3]
```diff
@@ -109,24 +109,29 @@ void RouteSync::onMsg(int nlmsg_type, struct nl_object *obj)
         /* 0 means no ifindex */
         if (m_link_cache)
         {
-            rtnl_link_i2name(m_link_cache, ifindex, addrStr, MAX_ADDR_SIZE);
-            ifindexes += addrStr;
+            rtnl_link_i2name(m_link_cache, ifindex, ifname, MAX_ADDR_SIZE);
+            if (!strlen(ifname))
+            {
+                rtnl_link_alloc_cache(m_nl_sock, AF_UNSPEC, &m_link_cache);
+                rtnl_link_i2name(m_link_cache, ifindex, ifname, MAX_ADDR_SIZE);
```
Comment (lguohan) [inline fpmsyncd/routesync.cpp]: could the ifname still not get resolved even after updating the cache? 
should we throw exception or print here if it happens?

File: fpmsyncd/routesync.cpp [hunk 4]
```diff
@@ -109,24 +109,29 @@ void RouteSync::onMsg(int nlmsg_type, struct nl_object *obj)
...
             nexthops += string(",");
-            ifindexes += string(",");
+            ifnames += string(",");
         }
     }
 
     std::vector<FieldValueTuple> fvVector;
     FieldValueTuple nh("nexthop", nexthops);
-    FieldValueTuple idx("ifindex", ifindexes);
+    FieldValueTuple idx("ifindex", ifnames);
```
Comment (lguohan) [inline fpmsyncd/routesync.cpp]: should we change this to ifname as well?

File: fpmsyncd/routesync.cpp (hunk 3 above)
Comment (stcheng) [inline fpmsyncd/routesync.cpp]: the ifname is currently only used as information. the orchagent doesn't depend on this value. i think for here, if the ifname cannot be resolved after updating the cache, we could just add "unknown" here.

Comment (stcheng) [inline fpmsyncd/routesync.cpp]: @lguohan

**Comment by stcheng** [comment]
🆙

**Comment by lguohan** [comment]
Ack

**Pull request merged**